dot/state/storage.go

```diff
@@ -190,13 +199,31 @@ func (s *StorageState) LoadFromDB(root common.Hash) (*trie.Trie, error) {
 		return nil, err
 	}
 
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	s.tries[t.MustHash()] = t
+	_, _ = s.tries.LoadOrStore(t.MustHash(), t)
 	return t, nil
 }
 
+func (s *StorageState) loadTrie(root *common.Hash) (*trie.Trie, error) {
+	if root == nil {
+		sr, err := s.blockState.BestBlockStateRoot()
+		if err != nil {
+			return nil, err
+		}
+		root = &sr
+	}
+
+	if t, has := s.tries.Load(*root); !has && t != nil {
```

Comment:
should this be `has && t != nil` not `!has`?

Reply:
good catch, updated!